src/tools/upgrade-factory-snapshot.c

```diff
@@ -0,0 +1,725 @@
+            SFMF_DEBUG("Running '%s' (queue=%s, pos=%d/%d, checked=%d): '%s'\n",
+                    task->name, queue->name, queue->current, queue->total, task->checked, cmds);
+            g_free(cmds);
+
+            if (async) {
+                deploy_task_queue_run_async(queue);
+            } else {
+                deploy_task_queue_run_sync(queue);
+            }
+
+            result = 1;
+        } else {
+            if (queue->finished_callback) {
+                queue->finished_callback(queue, queue->callback_user_data);
+            }
+
+            queue->finished = 1;
+        }
+    }
+
+    return result;
+}
+
+void deploy_task_queue_abort(struct DeployTaskQueue *queue)
+{
+    SFMF_LOG("Aborting queue: %s\n", queue->name);
+    queue->finished = 1;
+}
+
+gboolean upgrade_factory_snapshot_do_next_entry_cb(gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+    deploy_task_queue_next(ufs->deploy_queue, 1);
+    return FALSE;
+}
+
+gboolean upgrade_factory_snapshot_delayed_finish_cb(gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+    upgrade_factory_snapshot_set_running(ufs, FALSE);
+    upgrade_factory_snapshot_schedule_quit(ufs);
+    return FALSE;
+}
+
+void upgrade_factory_snapshot_finished_cb(struct DeployTaskQueue *queue, void *user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    // Reset queue in case we need to re-start it later
+    queue->current = -1;
+    g_timeout_add(1000, upgrade_factory_snapshot_delayed_finish_cb, ufs);
+}
+
+void upgrade_factory_snapshot_task_done_cb(struct DeployTaskQueue *queue, void *user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+    g_idle_add(upgrade_factory_snapshot_do_next_entry_cb, ufs);
+}
+
+void upgrade_factory_snapshot_broadcast_status(struct UpgradeFactorySnapshot *ufs,
+        const char *partition, int progress, const char *message)
+{
+    struct DeployTaskQueue *queue = ufs->deploy_queue;
+    struct DeployTask *task = deploy_task_queue_get_current_task(queue);
+
+    if (!task) {
+        SFMF_FAIL("Current task is invalid\n");
+    }
+
+    if (ufs->status.partition) {
+        g_free(ufs->status.partition);
+    }
+    ufs->status.partition = g_strdup(partition);
+
+    ufs->status.progress = progress;
+
+    if (ufs->status.message) {
+        g_free(ufs->status.message);
+    }
+    ufs->status.message = g_strdup(message);
+
+    int partition_current = 0;
+    int partition_total = ufs->partitions_length;
+    while (partition_current < partition_total) {
+        if (strcmp(partition, ufs->partitions[partition_current]) == 0) {
+            break;
+        }
+        partition_current++;
+    }
+
+    if (partition_current == partition_total) {
+        if (strcmp(partition, "") != 0) {
+            SFMF_WARN("Unknown partition: %s\n", partition);
+        }
+        partition_current = 0;
+        partition_total = 1;
+    }
```

Comment:
What happens here?

If ufs->status.partition_total is zero on entry, it will be 1 on exit.